File: Chapitre4/exercice4.2.py
import os
import sys

def main():
    # Taille du tampon (non strictement nécessaire en Python mais gardé pour la cohérence)
    BUFFER_SIZE = 256

    # Création des pipes (tube1: fils -> père, tube2: père -> fils)
    try:
        r1, w1 = os.pipe()
        r2, w2 = os.pipe()
    except OSError:
        print("Erreur lors de la création des pipes")
        sys.exit(1)

    try:
        pid = os.fork()
    except OSError:
        print("Erreur lors du fork")
        sys.exit(1)

    if pid == 0:  # --- PROCESSUS FILS ---
        os.close(r1)  # Ferme lecture tube1
        os.close(w2)  # Ferme écriture tube2

        # Sauvegarder le vrai stdout (terminal)
        stdout_terminal = os.dup(sys.stdout.fileno())

        # Saisie du mot
        # On utilise sys.stderr ou on écrit directement pour éviter les buffers
        sys.stdout.write("Saisissez le mot : ")
        sys.stdout.flush()
        mot = sys.stdin.readline().strip().split()[0] # Équivalent scanf %s

        # Rediriger stdout vers le tube1
        os.dup2(w1, sys.stdout.fileno())
        os.close(w1)

        # Envoyer le mot (le print va dans le tube maintenant)
        print(mot, end='')
        sys.stdout.flush()
        
        # Fermer stdout pour signaler EOF au père
        os.close(sys.stdout.fileno())

        # Restauration de stdout vers le terminal
        os.dup2(stdout_terminal, sys.stdout.fileno())
        os.close(stdout_terminal)

        # Lire le résultat du père depuis tube2
        # os.read renvoie des bytes, on utilise l'unpacking pour l'entier
        import struct
        data = os.read(r2, 4) # sizeof(int) est généralement 4
        os.close(r2)

        if data:
            valR = struct.unpack('i', data)[0]
            if valR == 1:
                print(f"Le mot '{mot}' existe dans le fichier.")
            else:
                print(f"Le mot '{mot}' n'existe pas dans le fichier.")

    else:  # --- PROCESSUS PÈRE ---
        os.close(w1)  # Ferme écriture tube1
        os.close(r2)  # Ferme lecture tube2

        # Rediriger stdin pour lire depuis le tube1
        os.dup2(r1, sys.stdin.fileno())
        os.close(r1)

        # Recevoir le mot du fils (via stdin redirigé)
        mot = sys.stdin.read().strip()

        valW = 0
        try:
            with open("./fichier.txt", "r") as fichier:
                for ligne in fichier:
                    if mot in ligne:
                        valW = 1
                        break
        except FileNotFoundError:
            print("Erreur : fichier.txt introuvable")
            sys.exit(1)

        # Envoyer le résultat au fils via tube2
        import struct
        os.write(w2, struct.pack('i', valW))
        os.close(w2)

        # Attendre la fin du fils
        os.wait()

if __name__ == "__main__":
    main()

File: Chapitre4/exercice4.2.c
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <fcntl.h>
#include <sys/types.h>
#include <sys/wait.h>
#include <string.h>

#define BUFFER_SIZE 256

int main() {
    int tube1[2];  // fils → père (le mot)
    int tube2[2];  // père → fils (le résultat)

    if (pipe(tube1) == -1 || pipe(tube2) == -1) {
        perror("pipe");
        exit(1);
    }

    pid_t fils = fork();
    if (fils == -1) { perror("fork"); exit(1); }

    if (fils == 0) {  
        close(tube1[0]);
        close(tube2[1]);

        //Sauvegarder le vrai stdout (terminal)
        int stdout_terminal = dup(STDOUT_FILENO);

        char mot[BUFFER_SIZE];
        printf("Saisissez le mot : ");
        fflush(stdout);
        scanf("%s", mot);

        // ── Envoyer le mot au père via tube
        dup2(tube1[1], STDOUT_FILENO);  // stdout → tube1
        close(tube1[1]);

        printf("%s", mot);
        fflush(stdout);
        close(STDOUT_FILENO);           // signaler EOF au père

        // restauration de stdout stdout vers le terminal
        dup2(stdout_terminal, STDOUT_FILENO);
        close(stdout_terminal);

        int valR;
        read(tube2[0], &valR, sizeof(int));
        close(tube2[0]);

        if (valR == 1)
            printf("Le mot '%s' existe dans le fichier.\n", mot);
        else
            printf("Le mot '%s' n'existe pas dans le fichier.\n", mot);

    } else { 
        close(tube1[1]);
        close(tube2[0]);

        // ── Recevoir le mot du fils via tube1
        dup2(tube1[0], STDIN_FILENO);   // stdin ← tube1
        close(tube1[0]);

        char mot[BUFFER_SIZE];
        scanf("%s", mot);          

        FILE *fichier = fopen("./fichier.txt", "r");
        if (!fichier) { perror("fopen"); exit(1); }

        char ligne[BUFFER_SIZE];
        int valW = 0;

        while (fgets(ligne, sizeof(ligne), fichier) != NULL) {
            if (strstr(ligne, mot) != NULL) {
                valW = 1;
                break;
            }
        }
        fclose(fichier);

        write(tube2[1], &valW, sizeof(int));
        close(tube2[1]);

        wait(NULL);
    }

    return 0;
}